File: go-algorithm/leetcode/problem352.go
package main

// 352. Data Stream as Disjoint Intervals
// https://leetcode.com/problems/data-stream-as-disjoint-intervals/

type SummaryRanges struct {
	Intervals [][]int
}

func _() SummaryRanges {
	return SummaryRanges{}
}

func (r *SummaryRanges) AddNum(val int) {
	if len(r.Intervals) == 0 {
		r.Intervals = append(r.Intervals, []int{val, val})
		return
	}
	for i, interval := range r.Intervals {
		if interval[0] <= val && val <= interval[1] {
			break
		}
		if interval[0] == val+1 {
			if i > 0 && r.Intervals[i-1][1] == val-1 {
				interval[0] = r.Intervals[i-1][0]
				r.Intervals = append(r.Intervals[:i-1], r.Intervals[i:]...)
				return
			} else {
				interval[0] = val
			}
			return
		}
		if interval[0] > val+1 {
			if i > 0 && r.Intervals[i-1][1]+1 == val {
				r.Intervals[i-1][1] = val
			} else {
				r.Intervals = append(r.Intervals[:i+1], r.Intervals[i:]...)
				r.Intervals[i] = []int{val, val}
			}
			return
		}
		if i == len(r.Intervals)-1 {
			if interval[1] == val-1 {
				interval[1] = val
			} else {
				r.Intervals = append(r.Intervals, []int{val, val})
			}
			return
		}
	}
}

func (r *SummaryRanges) GetIntervals() [][]int {
	return r.Intervals
}

File: python-algo/leetcode/problem_352.py
"""352. Data Stream as Disjoint Intervals
https://leetcode.com/problems/data-stream-as-disjoint-intervals/
"""
import bisect
from typing import List


class SummaryRanges:

    def __init__(self):
        self.intervals = []  # [[1,3], [5,5], [7,9]]
        self.order = []  # [3, 5, 9]

    def addNum(self, value: int) -> None:
        if not self.intervals:
            self.intervals.append([value, value])
            self.order.append(value)
            return
        pos = bisect.bisect_right(self.order, value)  # 向右插入的位置
        self.intervals.insert(pos, [value, value])
        self.order.insert(pos, value)
        deleted = set()
        if pos > 0:
            if self.intervals[pos][0] <= self.intervals[pos - 1][1]:  # [1,3], [3, 3]
                deleted.add(pos)
            elif self.intervals[pos][0] == self.intervals[pos - 1][1] + 1:  # [1,3], [4, 4]
                self.intervals[pos][0] = self.intervals[pos - 1][0]
                self.intervals[pos - 1] = self.intervals[pos]
                deleted.add(pos)
        if pos < len(self.intervals) - 1:
            if self.intervals[pos][1] + 1 == self.intervals[pos + 1][0]:
                self.intervals[pos][1] = self.intervals[pos + 1][1]
                self.intervals[pos + 1] = self.intervals[pos]
                deleted.add(pos + 1)
        new_intervals = []
        new_order = []
        for i, interval in enumerate(self.intervals):
            if i not in deleted:
                new_intervals.append(interval)
                new_order.append(self.order[i])
        self.intervals = new_intervals
        self.order = new_order

    def getIntervals(self) -> List[List[int]]:
        return self.intervals
